All_Relationships/9.entity_prediction.py

```diff
@@ -0,0 +1,230 @@
+
+# In[1]:
+
+
+get_ipython().magic(u'load_ext autoreload')
+get_ipython().magic(u'autoreload 2')
+get_ipython().magic(u'matplotlib inline')
+
+from collections import defaultdict
+import csv
+import os
+
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import tqdm
+from scipy.stats import fisher_exact
+import scipy
+from sqlalchemy import and_
+from sklearn.linear_model import LogisticRegression, SGDClassifier
+from sklearn.metrics import average_precision_score, precision_recall_curve, roc_curve, auc, f1_score, confusion_matrix
+from sklearn.model_selection import GridSearchCV
+import seaborn as sns
+
+
+# In[2]:
+
+
+candidate_df = pd.read_csv("disease_gene_summary_stats.csv")
+prior_df = pd.read_csv("observation-prior.csv")
+
+
+# In[3]:
+
+
+candidate_df.head(10)
+
+
+# In[4]:
+
+
+prior_df.head(10)
+
+
+# # Set up the Training and Testing Set
+
+# In[5]:
+
+
+train_df = pd.read_csv("stratified_data/training_set.csv")
+dev_df = pd.read_csv("stratified_data/dev_set.csv")
+test_df = pd.read_csv("stratified_data/test_set.csv")
+
+
+# In[6]:
+
+
+# Gather the summary stats for each candidate
+training_set = pd.merge(candidate_df, train_df, how='right', on=["disease_id", "gene_id"])
+dev_set = pd.merge(candidate_df, dev_df, how='right', on=["disease_id", "gene_id"])
+test_set = pd.merge(candidate_df, test_df, how='right', on=["disease_id", "gene_id"])
+
+
+# Drop the values that aren't found in pubmed. 
+training_set = training_set.drop("hetnet_labels", axis=1)
+dev_set = dev_set.drop("hetnet_labels", axis=1)
+test_set = test_set.drop("hetnet_labels", axis=1)
+
+training_set = training_set.dropna()
+dev_set = dev_set.dropna()
+test_set = test_set.dropna()
+
+# Add the prior prob to the different sets 
+training_set = pd.merge(training_set, prior_df[["disease_id", "gene_id", "prior_perm"]])
+dev_set = pd.merge(dev_set, prior_df[["disease_id", "gene_id", "prior_perm"]])
+test_set = pd.merge(test_set, prior_df[["disease_id", "gene_id", "prior_perm"]])
+
+
+# In[7]:
+
+
+non_features = ["hetnet","final_model_pred", "disease_id", "gene_id", "gene_name", "disease_name", "pubmed"]
+
+X = training_set[[col for col in training_set.columns if col not in non_features]]
+Y = training_set["hetnet"]
+
+X_dev = dev_set[[col for col in dev_set.columns if col not in non_features]]
+Y_dev = dev_set["hetnet"]
+
+X_test = test_set[[col for col in test_set.columns if col not in non_features]]
+Y_test = test_set["hetnet"]
+
+
+# # Train the Machine Learning Algorithms
+
+# Here we use gridsearch to optimize both models using 10 fold cross validation. After exhausting the list of parameters, the best model is chosen and analyzed in the next chunk. 
+
+# In[8]:
+
+
+n_iter = 100
+final_models = []
+
+lr = LogisticRegression()
+elastic_net = SGDClassifier(penalty='elasticnet', loss='log', random_state=100)
```

Comment:
Both `lr` and `elastic_net` are logistic regression. They seem redundant. We should get `elastic_net` to perform as good or better than `lr` and then drop `lr`?